### PR TITLE
Change USB filtering

### PR DESCRIPTION
**Pull request by dschuermann**
Instead of filtering by the USB xml, we will move to a method of filtering inside OpenKeychain, so we can display helpful messages to the user.

## Types of changes
- ✅ Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java [hunk 1]
```diff
@@ -211,36 +230,36 @@ public TransportType getTransportType() {
     @Override
     public TokenType getTokenType() {
         switch (usbDevice.getVendorId()) {
-            case 4176: {
+            case VID_YUBIKEY: {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java]: How about `VENDOR_` and `DEVICE_`? I find particularly `PID` misleading

Comment (dschuermann) [reply OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java]: 👍 I changed to ``VENDOR`` and ``PRODUCT`` as in the terminology of USB.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java [hunk 2]
```diff
@@ -202,6 +222,49 @@ public int hashCode() {
...
+                    case PRODUCT_NITROKEY_PRO:
+                        return TokenType.NITROKEY_PRO;
+                    case PRODUCT_NITROKEY_START:
+                        return TokenType.NITROKEY_START;
+                    case PRODUCT_NITROKEY_STORAGE:
+                        return TokenType.NITROKEY_STORAGE;
+                }
+                break;
+            }
+            case VENDOR_GNUK: {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java]: this is FSIJ, the Free Software Initiative Japan

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java [hunk 3]
```diff
@@ -202,6 +222,49 @@ public int hashCode() {
...
 
+    @Override
+    public TransportType getTransportType() {
+        return TransportType.USB;
+    }
+
+    @Override
+    public TokenType getTokenType() {
+        switch (usbDevice.getVendorId()) {
+            case VENDOR_YUBIKEY: {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/usb/UsbTransport.java]: `YUBICO` :)

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenHelper.java [hunk 4]
```diff
@@ -1003,9 +1005,26 @@ public SecurityTokenInfo getTokenInfo() throws IOException {
...
         byte[] pwInfo = getPwStatusBytes();
 
-        return SecurityTokenInfo.create(fingerprints, aid, userId, url, pwInfo[4], pwInfo[6]);
+        TransportType transportType = mTransport.getTransportType();
+        TokenType tokenType;
+        if (transportType == TransportType.USB) {
+            tokenType = mTransport.getTokenType();
+        } else {
+            tokenType = isFidesmoToken() ? TokenType.FIDESMO : TokenType.UNKNOWN_NFC;
+        }
```
Comment (dschuermann) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenHelper.java]: @Valodim I found a very strange problem here only happening for Fidesmo cards: In the current order of commands everything works, but if I first get the transport type including ``isFidesmoToken()`` and only afterwards get the fingerprints ``rawFingerprints`` will be null. Do you have an idea what could cause this?

**Comment by dschuermann** [comment]
@Valodim Also, please review.

Currently, the handling of not-supported USB devices is not perfect: I tried for example my smart card reader, which is also class 11, and it will error out much earlier inside UsbTransport.connect().

File: OpenKeychain/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -1580,6 +1580,8 @@
     <string name="security_token_error_header">"Security Token reported invalid %s byte."</string>
     <string name="security_token_error_tag_lost">"Security Token has been taken off too early. Keep the Security Token at the back until the operation finishes."</string>
     <string name="security_token_error_iso_dep_not_supported">"Security Token does not support the required communication standard (ISO-DEP, ISO 14443-4)"</string>
+    <string name="security_token_error_generic">"Generic Error: Most probably, the Security Token has been taken off too early."</string>
```
Comment (Valodim) [inline OpenKeychain/src/main/res/values/strings.xml]: "Generic" is a very technical term

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenInfo.java [hunk 6]
```diff
@@ -74,9 +80,41 @@ public static SecurityTokenInfo newInstanceDebugLockedHard() {
...
+        return SecurityTokenInfo.create(TransportType.NFC, TokenType.UNKNOWN_NFC,
                 new byte[][] { KeyFormattingUtils.convertFingerprintHexFingerprint("4700BA1AC417ABEF3CC7765AD686905837779C3E") },
                 Hex.decode("010203040506"), "yubinu2@mugenguild.com", "http://valodim.stratum0.net/mryubinu2.asc", 0, 0);
     }
 
+    public enum TransportType {
+        NFC, USB
+    }
+
+    public enum TokenType {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenInfo.java]: Is there a particular reason you put UNKNOWN_NFC and UNKNOWN_USB here redundantly? This information is already available from the TransportType.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenConnection.java [hunk 7]
```diff
@@ -948,12 +950,29 @@ public SecurityTokenInfo getTokenInfo() throws IOException {
         String url = getUrl();
         byte[] pwInfo = getPwStatusBytes();
 
-        return SecurityTokenInfo.create(fingerprints, aid, userId, url, pwInfo[4], pwInfo[6]);
+        TransportType transportType = mTransport.getTransportType();
```
Comment (dschuermann) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenConnection.java]: @Valodim I found a very strange problem here only happening for Fidesmo cards: In the current order of commands everything works, but if I first get the transport type including ``isFidesmoToken()`` and only afterwards get the fingerprints ``rawFingerprints`` will be null. Do you have an idea what could cause this?

**Comment by dschuermann** [comment]
@Valodim done

**Comment by Valodim** [comment]
I'm still unsure whether the fidesmo check is really necessary. But it's fine if it's in for now.

Also, I changed the "generic error" string to "communication error" on master.

**Pull request merged**